util/config.py

```diff
@@ -0,0 +1,432 @@
+                structpass.struct[curfield] = value
+            else:
+                structpass.struct = { curfield: value }
+        else:
+            structchuck = types.SimpleNamespace()
+            try:
+                nextifield = int( fields[1] )
+            except ValueError:
+                nextifield = None
+
+            if isinstance( structpass.struct, list ):
+                structchuck.struct = {} if nextifield is None else []
+                self.set_value( ".".join(fields[1:], value, structchuck, appendlists=appendlists ) )
+                if appendlists:
+                    if ifield is None:
+                        raise TypeError( "Tried to add a non-integer field to a list" )
+                    structpass.struct.append( structchuck.struct )
+                else:
+                    if ifield is None:
+                        structpass.struct = { curfield: structchuck.struct }
+                    else:
+                        structpass.struct = [ structchuck.struct ]
+            else:
+                if ifield is None:
+                    if isinstance( structpass.struct, dict ):
+                        if curfield in structpass.struct:
+                            structchuck.struct = structpass.struct[curfield]
+                        else:
+                            structchuck.struct = {} if nextifield is None else []
+                    else:
+                        structpass.struct = {}
+                    self.set_value( ".".join(fields[1:]), value, structchuck, appendlists=appendlists )
+                    structpass.struct[curfield] = structchuck.struct
+                else:
+                    if isinstance( structpass.struct, dict ):
+                        raise TypeError( "Tried to add an integer field to a dict" )
+                    structchuck.struct = {} if nextifield is None else []
+                    self.set_value( ".".join(fields[1:]), value, structchuck, appendlists=appendlists )
+                    structpass.struct = [ structchuck.struct ]
+
+    @classmethod
+    def _fieldsep( cls, field ):
+        """Parses a period-separated config specifier string.
+        Returns:
+          fields : list of the hierarchy (e.g. "val1.val2.val3" returns ["val1","val2","val3"])
+          isleaf : True if len(fields) is 1, otherwise false
+          curfield : The first element of the field (val1 in the example above)
+          ifield : None if curfield is not an integer, otherwise the integer value of curfield
+        """
+        fields = field.split( "." )
+        isleaf = ( len(fields) == 1 )
+        curfield = fields[0]
+        try:
+            ifield = int(curfield)
+        except ValueError:
+            ifield = None
+        return fields, isleaf, curfield, ifield
+
+    @staticmethod
+    def merge_trees( left, right, augment=False ):
+        """Internal usage, do not call."""
```

Comment:
I have come to appreciate that even internal functions should have a proper docstring. Down the line it could help make things clearer when trying to debug something. 

I would prefer if we use "numpy" style for functions:
```
"""
Function description...

Parameters
--------------
par1: type
  description
par2: type
  description

Returns
----------
return1: type
  description
```

Reply:
Yeah, I agree that internal functions should have a docstring.  I was just lazy.  (But, I also think it's worth keeping the "internal function" warning in the docstring.  Occasionally it's an advantage of Python that it doesn't have private methods like C++ or Java, but there is value in posting a warning that you shouldn't use something externally.)

Reply:
Yeah, another option is to change the function name to start with an underscore.